src/Aspire.Hosting/Dashboard/DashboardServiceData.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Dcp;
+using Microsoft.Extensions.Logging;
+
+namespace Aspire.Hosting.Dashboard;
+
+/// <summary>
+/// Models the state for <see cref="DashboardService"/>, as that service is constructed
+/// for each gRPC request. This long-lived object holds state across requests.
+/// </summary>
+internal sealed class DashboardServiceData : IAsyncDisposable
+{
+    private readonly CancellationTokenSource _cts = new();
+    private readonly ResourcePublisher _resourcePublisher;
+    private readonly ConsoleLogPublisher _consoleLogPublisher;
+
+    public DashboardServiceData(
+        DistributedApplicationModel applicationModel,
+        KubernetesService kubernetesService,
+        ILoggerFactory loggerFactory)
+    {
+        _resourcePublisher = new ResourcePublisher(_cts.Token);
+        _consoleLogPublisher = new ConsoleLogPublisher(_resourcePublisher);
+
+        _ = new DcpDataSource(kubernetesService, applicationModel, loggerFactory, _resourcePublisher.IntegrateAsync, _cts.Token);
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        await _cts.CancelAsync().ConfigureAwait(false);
+
+        _cts.Dispose();
+    }
+
+    internal ResourceSnapshotSubscription SubscribeResources()
+    {
+        return _resourcePublisher.Subscribe();
+    }
+
+    internal IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>>? SubscribeConsoleLogs(string resourceName)
+    {
+        var sequence = _consoleLogPublisher.Subscribe(resourceName);
+
+        return sequence is null ? null : Enumerate();
+
+        async IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>> Enumerate()
```

Comment:
nit: I prefer static local methods. Could pass in the cts and sequence as parameters and make this static.

Reply:
I use both static and non-static local methods on a case-by-case basis.